brave/src/main/java/com/linecorp/armeria/client/brave/ClientRequestContextAdapter.java

```diff
@@ -0,0 +1,179 @@
+     */
+    static brave.http.HttpClientRequest asHttpClientRequest(ClientRequestContext ctx,
+        RequestHeadersBuilder headersBuilder) {
+        return new HttpClientRequest(ctx, headersBuilder);
+    }
+
+    private static final class HttpClientRequest extends brave.http.HttpClientRequest {
+        private final ClientRequestContext ctx;
+        private final RequestHeadersBuilder headersBuilder;
+
+        HttpClientRequest(ClientRequestContext ctx, RequestHeadersBuilder headersBuilder) {
+            this.ctx = ctx;
+            this.headersBuilder = headersBuilder;
+        }
+
+        @Override
+        public ClientRequestContext unwrap() {
+            return ctx;
+        }
+
+        @Override
+        public String method() {
+            return ctx.method().name();
+        }
+
+        /**
+         * Original implementation is calling {@link HttpClientAdapter#url(Object)} which needs
+         * {@link RequestLog#scheme()} is not available at {@link RequestLogAvailability#REQUEST_START}.
+         * We need to use {@link RequestLog#path()} directly.
+         *
+         * @see brave.http.HttpClientRequest#path()
+         */
+        @Override
+        public String path() {
+            return ctx.path();
+        }
+
+        @Override
+        @Nullable
+        public String url() {
+            return SpanTags.generateUrl(ctx.log());
+        }
+
+        @Override
+        @Nullable
+        public String header(String name) {
+            if (!ctx.log().isAvailable(RequestLogAvailability.REQUEST_HEADERS)) {
+                return null;
+            }
+            return ctx.log().requestHeaders().get(name);
+        }
+
+        @Override
+        public long startTimestamp() {
+            if (!ctx.log().isAvailable(RequestLogAvailability.REQUEST_START)) {
+                return 0L;
+            }
+            return ctx.log().requestStartTimeMicros();
+        }
+
+        @Override
+        public void header(String name, String value) {
+            headersBuilder.set(name, value);
+        }
+    }
+
+    static brave.http.HttpClientResponse asHttpClientResponse(ClientRequestContext ctx) {
+        return new HttpClientResponse(ctx);
+    }
+
+    private static final class HttpClientResponse extends brave.http.HttpClientResponse {
+        private final ClientRequestContext ctx;
+
+        HttpClientResponse(ClientRequestContext ctx) {
+            this.ctx = ctx;
+        }
+
+        @Override
+        public ClientRequestContext unwrap() {
+            return ctx;
+        }
+
+        @Override
+        public int statusCode() {
+            if (!ctx.log().isAvailable(RequestLogAvailability.RESPONSE_HEADERS)) {
+                return 0;
+            }
+            return ctx.log().status().code();
+        }
+
+        @Override
+        public long finishTimestamp() {
+            return SpanContextUtil.wallTimeMicros(ctx.log(), ctx.log().responseEndTimeNanos());
```

Comment:
Do we need a check here?
```java
if (!ctx.log().isAvailable(RequestLogAvailability.RESPONSE_END)) {
    return 0L;
}
```

Reply:
note for future... this all is terribly crufty. it is sad that these operations cannot return absent values

Reply:
😭

Reply:
thanks regardless. I'd hate for some accidental usage of this to crash a request. In hind sight, probably this method should have been protected.

Reply:
https://github.com/line/armeria/issues/2090 on my complaint :) thanks for being watchdog meanwhile. If this is solved, you and @kojilin etc can be less scared of api usage as it doesn't need to be so scary anyway